### PR TITLE
Add support for secondaryKeys and default factory

### PR DESCRIPTION
**Pull request by thoov**
- [x] tests

### CODE REVIEW

File: addon/state-for.js [hunk 1]
```diff
@@ -60,25 +71,36 @@ export default function stateFor(stateName, propertyName) {
...
-function createStateFor(context, stateName, container) {
+function createStateFor(context, stateName) {
   let defaultState  = {};
   let containerName = `state:${stateName}`;
-  let StateFactory  = container.lookupFactory(containerName);
+  let StateFactory  = context.container.lookupFactory(containerName);
 
   if (!StateFactory) {
-    throw new TypeError(`Unknown StateFactory: ${containerName}`);
+    return Object.create(null); // default to a blank object if no factory was found.
```
Comment (stefanpenner) [inline addon/state-for.js]: We should actually create an empty object. As they allocate quite abit faster

File: addon/state-for.js [hunk 2]
```diff
@@ -48,11 +57,13 @@ export default function stateFor(stateName, propertyName) {
     let state = weakMaps[stateName];
 
     if (!state.has(propertyValue)) {
-      let newState = createStateFor(this, stateName, this.container);
-      state.set(propertyValue, newState);
+      let secondaryMap = Ember.Map.create();
```
Comment (stefanpenner) [inline addon/state-for.js]: As the keys are always strings Thoughts on this just being a an object?

**Comment by thoov** [comment]
@stefanpenner can you take another look. I removed guidFor as its not what we need and used btoa.

File: addon/state-for.js [hunk 3]
```diff
@@ -48,11 +56,15 @@ export default function stateFor(stateName, propertyName) {
     let state = weakMaps[stateName];
 
     if (!state.has(propertyValue)) {
-      let newState = createStateFor(this, stateName, this.container);
-      state.set(propertyValue, newState);
+      state.set(propertyValue, {
+        [guidString(secondaryKeys)]: createStateFor(this, stateName)
```
Comment (stefanpenner) [inline addon/state-for.js]: seems like we could `let guid = guidString(secondaryKeys)` and use `guid` throughout the function.

File: addon/state-for.js [hunk 4]
```diff
@@ -48,11 +56,15 @@ export default function stateFor(stateName, propertyName) {
...
+      });
+    }
+    else if(!state.get(propertyValue)[guidString(secondaryKeys)]) {
```
Comment (stefanpenner) [inline addon/state-for.js]: ^

File: addon/state-for.js [hunk 5]
```diff
@@ -48,11 +56,15 @@ export default function stateFor(stateName, propertyName) {
...
+      state.get(propertyValue)[guidString(secondaryKeys)] = createStateFor(this, stateName);
```
Comment (stefanpenner) [inline addon/state-for.js]: ^

File: addon/state-for.js [hunk 6]
```diff
@@ -48,11 +56,15 @@ export default function stateFor(stateName, propertyName) {
...
     }
 
-    return state.get(propertyValue);
+    return state.get(propertyValue)[guidString(secondaryKeys)];
```
Comment (stefanpenner) [inline addon/state-for.js]: ^

File: addon/state-for.js (hunk 5 above)
Comment (stefanpenner) [inline addon/state-for.js]: should as pass the secondary keys as a arg to the state factory?

File: addon/state-for.js (hunk 3 above)
Comment (thoov) [inline addon/state-for.js]: :+1:

File: addon/state-for.js [hunk 7]
```diff
@@ -3,42 +3,50 @@ import WeakMap from 'ember-weakmap/weak-map';
...
+* which is assoicated with the 'dependent-key-path' property. An optional third
+* argument can be passed to give a more granular division of state.
 *
-* stateFor('state-name').delete(this.get('property-name'));
-* stateFor('state-name').has(this.get('some-other-prop'));
-* stateFor('state-name').set(this.get('property-name'), {}); // Overrides the state
-* stateFor('state-name').get(this.get('property-name'));
+* stateFor('state-name', 'dependent-key-path');
+* stateFor('state-name', 'dependent-key-path',  {
+*    secondaryKeys: ['foo', 'bar']
```
Comment (stefanpenner) [inline addon/state-for.js]: so secondary keys are just namespaces right? Not bound keys.

Comment (thoov) [inline addon/state-for.js]: right now they are just namespaces. bound keys get weird if they are bound to non primitives as it was throwing guidFor off

Comment (thoov) [inline addon/state-for.js]: @stefanpenner i guess the question is do we want to mix bound/unbound keys in:
```js
{
  secondaryKeys: [...]
}

// vs

{
  secondaryKeys: [...],
  boundKeys: [...]
}
```

Comment (stefanpenner) [inline addon/state-for.js]: just spitballing but should we maybe call the `namespaces` or `scope` ?

Comment (thoov) [inline addon/state-for.js]: ya I like namespaces better

Comment (stefanpenner) [inline addon/state-for.js]: Ya we can punt on bound keys for now, (they make stuff tricky)

**Comment by thoov** [comment]
@stefanpenner this should be good to go. Maybe if you could do one more pass on it

**Pull request closed without merging**